### PR TITLE
994 detect collisions for voxel vs voxel and voxel vs box

### PR DESCRIPTION
**Pull request by joaomanita**
# Description
Added a voxel collision shape which decomposes a voxel grid into a set of boxes to be used in collision detection.
Added a greedy meshing like algorithm to do the decomposing into boxes.
Added systems to detect collisions between voxel collision shapes and box shapes, and voxel-voxel.

## Checklist

- [X] Self-review changes.
- [ ] Evaluate impact on the documentation.
- [ ] Ensure test coverage.
- [ ] Write new samples.
- [ ] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: engine/src/collisions/plugin.cpp [hunk 1]
```diff
@@ -42,4 +142,58 @@ void cubos::engine::collisionsPlugin(Cubos& cubos)
...
+                        }
+                        if (possibleSize.x > 0)
+                        {
+                            expandBoxYaxis(copiedGrid, currentMin, size, possibleSize);
+                            glm::vec3 boxCenter = glm::vec3(currentMin + static_cast<glm::ivec3>(possibleSize)) / 2.0F;
+                            cubos::core::geom::Box box;
+                            box.halfSize = boxCenter;
+                            glm::vec3 shift = center - boxCenter;
+                            shape.insertBox(box, shift);
+                            CUBOS_WARN("inserted pair");
```
Comment (luishfonseca) [inline engine/src/collisions/plugin.cpp:187]: was this forgotten here?

**Review by luishfonseca — DISMISSED** [review]
Needs a sample to be approved, I want to play a bit with it before.
The code overall looks quite nice, good job :+1:

File: engine/src/collisions/narrow_phase/plugin.cpp [hunk 2]
```diff
@@ -65,4 +66,86 @@ void cubos::engine::narrowPhaseCollisionsPlugin(Cubos& cubos)
...
+                        if (intersects)
+                        {
+                            cmds.relate(ent1, ent2,
+                                        CollidingWith{.entity = ent1,
+                                                      .entity1OriginalPosition = position1.vec,
+                                                      .entity2OriginalPosition = position2.vec,
+                                                      .penetration = intersectionInfo.penetration,
+                                                      .position = {0.0F, 0.0F, 0.0F},
+                                                      .normal = intersectionInfo.normal});
+                            break;
```
Comment (RiscadoA) [inline engine/src/collisions/narrow_phase/plugin.cpp:141]: This is still missing the multiple intersection handling code. Maybe we should leave it to another PR, what do you think?

**Review by RiscadoA — CHANGES_REQUESTED** [review]
Thanks for working on this! Also mostly looks good to me, just needs a bit clean up.

File: engine/src/collisions/narrow_phase/plugin.cpp [hunk 3]
```diff
@@ -113,78 +198,229 @@ void cubos::engine::narrowPhaseCollisionsPlugin(Cubos& cubos)
...
+                        // Create a translation matrix for the shift
+                        glm::mat4 shiftMatrix = glm::translate(glm::mat4(1.0F), -box.shift);
+                        matchedLocalToWorld2.mat *= shiftMatrix;
 
-                    // If we flipped incident and reference planes, we will
-                    // need to flip it back before sending it to the manifold.
-                    if (flipped)
+                        auto newPoints = computePoints(matchedShape1, matchedLocalToWorld1, matchedShape2,
+                                                       &matchedLocalToWorld2, collidingWith);
+                        points.insert(points.end(), newPoints.begin(), newPoints.end());
```
Comment (fallenatlas) [inline engine/src/collisions/narrow_phase/plugin.cpp:332]: Correct me if I'm wrong, here you are going through all boxes in the voxel shape and computing the manifold with the box, right? However, the collidingWith was already performed between each of these, which means that we should know exactly which boxes withing the voxel shape are colliding with this one. I think that we could maybe keep like an index or whatever is best of the specific box in the voxel shape that we collided with and only do computePoints between that and the box. 

Extra, but for more context: The computation of the manifold assumes the shapes are colliding, the only sanity check it performs is with penetration in the direction of the normal, and we are computing it between boxes that may not be colliding, I think this can create incorrect forces in the solver in some occasions.

**Review by fallenatlas — CHANGES_REQUESTED** [review]
Overall looks good to me, good job 👍 Just have a couple comments regarding the manifold part.

**Review by fallenatlas — COMMENTED** [review]
Another thing, in the voxel collision sample it would be nice to add the manifold visualization like we have in the other sample, would help to make sure everything is correct.

File: engine/src/collisions/narrow_phase/plugin.cpp (hunk 3 above)
Comment (fallenatlas) [reply engine/src/collisions/narrow_phase/plugin.cpp:332]: Disregard my comment here, I think it's something still worth thinking about but we have to change other things, before worrying about this.

File: engine/include/cubos/engine/collisions/shapes/voxel.hpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
+/// @file
+/// @brief Component @ref cubos::engine::VoxelCollisionShape.
+/// @ingroup collisions-plugin
+
+#pragma once
+
+#include <glm/glm.hpp>
+
+#include <cubos/core/geom/box.hpp>
+#include <cubos/core/reflection/external/glm.hpp>
```
Comment (RiscadoA) [inline engine/include/cubos/engine/collisions/shapes/voxel.hpp:10]: You only need this include in the .cpp where you define the reflection

File: engine/samples/voxel-shape-collisions/main.cpp [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            .add(Rotation::lookingAt({3.0F, -1.0F, 0.0F}, glm::vec3{0.0F, 1.0F, 0.0F}));
+    });
+
+    cubos.startupSystem("configure Assets").tagged(settingsTag).call([](Settings& settings) {
+        settings.setString("assets.io.path", SAMPLE_ASSETS_FOLDER);
+    });
+
+    /// [Set palette]
+    cubos.startupSystem("set palette").call([](RenderPalette& palette) { palette.asset = PaletteAsset; });
+    /// [Set palette]
```
Comment (RiscadoA) [inline engine/samples/voxel-shape-collisions/main.cpp:98]:
```suggestion
    cubos.startupSystem("set palette").call([](RenderPalette& palette) { palette.asset = PaletteAsset; });
```
You aren't using these, since the sample doesn't have its own page anyway.

File: engine/samples/voxel-shape-collisions/main.cpp [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    Entity b;
+
+    glm::vec3 aRotationAxis;
+    glm::vec3 bRotationAxis;
+};
+
+/// [Get handles to assets]
+static const Asset<VoxelGrid> CarAsset = AnyAsset("059c16e7-a439-44c7-9bdc-6e069dba0c75");
+static const Asset<VoxelPalette> PaletteAsset = AnyAsset("1aa5e234-28cb-4386-99b4-39386b0fc215");
+/// [Get handles to assets]
```
Comment (RiscadoA) [inline engine/samples/voxel-shape-collisions/main.cpp:69]:
```suggestion
static const Asset<VoxelGrid> CarAsset = AnyAsset("059c16e7-a439-44c7-9bdc-6e069dba0c75");
static const Asset<VoxelPalette> PaletteAsset = AnyAsset("1aa5e234-28cb-4386-99b4-39386b0fc215");
```

File: engine/src/collisions/plugin.cpp [hunk 7]
```diff
@@ -1,19 +1,193 @@
...
+            }
+        }
+    }
+    max = glm::ivec3(max.x + 1, max.y + 1, max.z + 1);
+    glm::vec3 dist = glm::vec3(max) - glm::vec3(min);
+    aabb.min({-dist.x / 2.0F, -dist.y / 2.0F, -dist.z / 2.0F});
+    aabb.max({dist.x / 2.0F, dist.y / 2.0F, dist.z / 2.0F});
+}
+
+pair<glm::uvec3, glm::uvec3> findLargestBox(const cubos::engine::VoxelGrid& grid,
```
Comment (RiscadoA) [inline engine/src/collisions/plugin.cpp:48]:
```suggestion
static pair<glm::uvec3, glm::uvec3> findLargestBox(const cubos::engine::VoxelGrid& grid,
```

File: engine/src/collisions/plugin.cpp [hunk 8]
```diff
@@ -1,19 +1,193 @@
...
+            {
+                processed[x][y][z] = true;
+            }
+        }
+    }
+
+    return std::make_pair(start, end);
+}
+
+vector<pair<glm::uvec3, glm::uvec3>> greedy3dMeshing(const cubos::engine::VoxelGrid& grid)
```
Comment (RiscadoA) [inline engine/src/collisions/plugin.cpp:160]:
```suggestion
static vector<pair<glm::uvec3, glm::uvec3>> greedy3dMeshing(const cubos::engine::VoxelGrid& grid)
```

File: engine/tests/settings.cpp [hunk 9]
```diff
@@ -52,12 +52,21 @@ TEST_CASE("cubos::engine::settingsPlugin")
     };
 
     EmbeddedArchive::registerData("settings.json", settingsData);
-    FileSystem::mount("/settings.json", std::make_unique<EmbeddedArchive>("settings.json"));
+
+    if (!FileSystem::mount("/settings.json", std::make_unique<EmbeddedArchive>("settings.json")))
+    {
+        FileSystem::unmount("/settings.json");
+        FileSystem::mount("/settings.json", std::make_unique<EmbeddedArchive>("settings.json"));
+    }
```
Comment (RiscadoA) [inline engine/tests/settings.cpp:60]: ? what is going on here

File: engine/tests/settings.cpp [hunk 10]
```diff
@@ -52,12 +52,21 @@ TEST_CASE("cubos::engine::settingsPlugin")
...
 
     // Run the actual test.
     Cubos cubos{argc, argv};
     cubos.plugin(settingsPlugin);
 
+    cubos.startupSystem("configure Assets").tagged(settingsTag).call([](Settings& settings) {
+        settings.setBool("assets.io.enable", false);
+    });
```
Comment (RiscadoA) [inline engine/tests/settings.cpp:68]: Why is this here?

**Review by RiscadoA — CHANGES_REQUESTED** [review]

**Review by RiscadoA — CHANGES_REQUESTED** [review]
You still haven't resolved my new comments :thinking:

File: engine/tests/settings.cpp (hunk 9 above)
Comment (joaomanita) [reply engine/tests/settings.cpp:60]: This test was failing because "/settings.json" was already mounted from executing the raycast test, getting a "could not mount archive error". I don't really know how the file system works so I just unmounted and remounted.

**Comment by joaomanita** [comment]
> You still haven't resolved my new comments 🤔

Oops, I re-requested a review without noticing you had already reviewed it :innocent:

File: engine/include/cubos/engine/collisions/shapes/voxel.hpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+#include <cubos/engine/assets/asset.hpp>
+#include <cubos/engine/voxels/grid.hpp>
+
+namespace cubos::engine
+{
+    /// @brief Component which adds a voxel collision shape to an entity, used with a @ref Collider component.
+    /// @ingroup collisions-plugin
+    class VoxelCollisionShape
+    {
+
```
[the remark this answers is not in the document]
Comment (RiscadoA) [reply engine/include/cubos/engine/collisions/shapes/voxel.hpp:23]: Don't forget this!

File: engine/tests/settings.cpp (hunk 10 above)
Comment (RiscadoA) [reply engine/tests/settings.cpp:68]: Don't mark stuff as resolved without answering them! Otherwise I might just forget about it.
This test doesn't even use assets so it really shouldn't need this

**Review by RiscadoA — CHANGES_REQUESTED** [review]

**Review by RiscadoA — APPROVED** [review]
From my side, LGTM! Regarding the actual mechanics and design, I'll leave that to the physics team members which are way more knowledgeable in this than I am.

**Comment by RiscadoA** [comment]
Don't forget to add an entry to the changelog @joaomanita!

**Review by GCeSilva — APPROVED** [review]
LGTM!
